File: plotlykt-core/src/jvmMain/resources/js/plotlyConnect.js
/**
 * Use existing plotly or load it from the CDN if it is not available
 * @param action
 */
function withPlotly(action) {
    if (typeof Plotly !== "undefined") {
        action(Plotly);
    } else if (typeof window.promiseOfPlotly !== "undefined") {
        window.promiseOfPlotly.then(plotly => action(plotly));
    } else {
        console.warn("Plotly not defined. Loading the script from CDN")
        window.promiseOfPlotly = new Promise((accept, reject) => {
            let plotlyLoaderScript = document.createElement("script");
            plotlyLoaderScript.src = "https://cdnjs.cloudflare.com/ajax/libs/plotly.js/1.54.6/plotly.min.js";
            plotlyLoaderScript.type = 'text/javascript';
            plotlyLoaderScript.onload = () => {
                accept(Plotly);
            }
            plotlyLoaderScript.onerror = (error) => {
                console.error(error);
                reject(error)
            }
            document.head.appendChild(plotlyLoaderScript);
        });
    }
}

/**
 * Request and parse json from given address
 * @param url {URL}
 * @param callback
 * @return Promise<Json>
 */
function getJSON(url, callback) {

    function handleErrors(response) {
        if (!response.ok) {
            throw Error(response.statusText);
        }
        return response;
    }

    try {
        fetch(url, {
            method: 'GET',
            headers: {
                Accept: 'application/json',
            }
        })
            .then(handleErrors)
            .then(response => response.json())
            .then(json => callback(json))
            .catch(error => console.log(error));
    } catch (e) {
        alert("Fetch of plot data failed with error: " + e)
    }
}

/**
 * Safe call for Plotly.newPlot
 * @param id
 * @param data
 * @param layout
 * @param config
 */
function makePlot(id, data, layout, config) {
    withPlotly(plotly => plotly.newPlot(id, data, layout, config))
}

/**
 * Create a plot taking data from given url
 * @param id {string} element id for plot
 * @param from {URL} json server url
 * @param config {object} plotly configuration
 * @return {JSON}
 */
function createPlotFrom(id, from, config = {}) {
    getJSON(from, json => withPlotly(plotly => {
        plotly.newPlot(id, json.data, json.layout, config)
    }));
}

/**
 * Update a plot taking data from given url
 * @param id {string} element id for plot
 * @param from {URL} json server url
 * @return {JSON}
 */
function updatePlotFrom(id, from) {
    getJSON(from, json => withPlotly(plotly => plotly.react(id, json.data, json.layout)));
}

/**
 * Start pull updates with regular requests from client side
 * @param id {string}
 * @param from
 * @param millis
 */
function startPull(id, from, millis) {
    let action = function () {
        updatePlotFrom(id, from)
    };
    window.setInterval(action, millis)
}

/**
 * Start push updates via websocket
 * @param id {string} element id for plot
 * @param ws {URL} a websocket address
 */
function startPush(id, ws) {
    let socket = new WebSocket(ws);

    socket.onopen = function () {
        console.log("[Plotly.kt] A connection for plot with id = " + id + " with server established on " + ws);
    };

    socket.onclose = function (event) {
        if (event.wasClean) {
            console.log("The connection with server is closed");
        } else {
            console.log("The connection with server is broken"); // Server process is dead
        }
        console.log('Code: ' + event.code + ' case: ' + event.reason);
    };

    socket.onerror = function (error) {
        console.error("Ploty push update error: " + error.message);
        socket.close()
    };

    socket.onmessage = function (event) {
        //console.log('got message: ' + event.data);
        let json = JSON.parse(event.data);
        //TODO check if plotly is initialized in a cell
        if (json.plotId === id) {
            if (json.contentType === "layout") {
                withPlotly(plotly => plotly.relayout(id, json.content));
            } else if (json.contentType === "trace") {
                let content = json.content;
                //This is done to satisfy plotly requirements of arrays-in-arrays for data
                if (content.hasOwnProperty('x')) {
                    content.x = [content.x];
                }
                if (content.hasOwnProperty('y')) {
                    content.y = [content.y];
                }
                if (content.hasOwnProperty('z')) {
                    content.z = [content.z];
                }
                withPlotly(plotly => plotly.restyle(id, content, json['trace']));
            }
        }
    };

    //gracefully close socket just in case
    window.onbeforeunload = function () {
        console.log("Gracefully closing socket");
        socket.onclose = function () {
        }; // disable onclose handler first
        socket.close();
    };
}

File: plotlykt-core/src/jsMain/kotlin/space/kscience/plotly/plotlyConnect.kt
package space.kscience.plotly

import kotlinx.browser.document
import kotlinx.browser.window
import org.w3c.dom.CloseEvent
import org.w3c.dom.ErrorEvent
import org.w3c.dom.HTMLElement
import org.w3c.dom.WebSocket
import org.w3c.dom.url.URL
import org.w3c.fetch.RequestInit
import kotlin.js.json


/**
 * Wait for Plotly library to be loaded
 */
private fun withPlotly(block: PlotlyJs.() -> Unit) {
    if (jsTypeOf(PlotlyJs) != "undefined"){ //if already defined, use it
        PlotlyJs.block()
    } else {
        //TODO add fetch
        error("Plotly is not defined")
    }
}

/**
 * Request and parse json from given address
 * @param url {URL}
 * @param callback
 * @return Promise<Json>
 */
private fun getJSON(url: URL, callback: (dynamic) -> Unit) {
    try {
        window.fetch(
            url,
            RequestInit(
                method = "GET",
                headers = json("Accept" to "application/json")
            )
        ).then { response ->
            if (!response.ok) {
                error("Fetch request failed with error: ${response.statusText}")
            } else {
                response.json().then(callback)
            }
        }.catch { error -> console.log(error) }
    } catch (e: Exception) {
        window.alert("Fetch of plot data failed with error: $e")
    }
}

/**
 * Create a plot taking data from given url
 * @param id {string} element id for plot
 * @param from {URL} json server url
 * @param config {object} plotly configuration
 * @return {JSON}
 */
@JsExport
public fun createPlotFrom(id: String, from: URL, config: dynamic = {}) {
    getJSON(from) { json ->
        val element = document.getElementById(id) as HTMLElement
        withPlotly { newPlot(element, json.data, json.layout, config) }
    }
}

/**
 * Update a plot taking data from given url
 * @param id {string} element id for plot
 * @param from {URL} json server url
 * @return {JSON}
 */
@JsExport
public fun updatePlotFrom(id: String, from: URL) {
    getJSON(from) { json ->
        val element = document.getElementById(id) as HTMLElement
        withPlotly { react(element, json.data, json.layout) }
    }
}

/**
 * Start pull updates with regular requests from client side
 * @param id {string}
 * @param from
 * @param millis
 */
@JsExport
public fun startPull(id: String, from: URL, millis: Int) {
    window.setInterval({ updatePlotFrom(id, from) }, millis)
}

/**
 * Start push updates via websocket
 * @param id {string} element id for plot
 * @param ws {URL} a websocket address
 */
@JsExport
public fun startPush(id: String, ws: String) {
    val element = document.getElementById(id) as HTMLElement
    val socket = WebSocket(ws)

    socket.onopen = {
        console.log("[Plotly.kt] A connection for plot with id = $id with server established on $ws")
    }

    socket.onclose = { event ->
        event as CloseEvent
        if (event.wasClean) {
            console.log("The connection with server is closed")
        } else {
            console.log("The connection with server is broken") // Server process is dead
        }
        console.log("Code: ${event.code} case: ${event.reason}")
    }

    socket.onerror = { error ->
        error as ErrorEvent
        console.error("Plotly push update error: " + error.message)
        socket.close()
    }

    socket.onmessage = { event ->
        val json: dynamic = JSON.parse(event.data.toString())
        if (json.plotId === id) {
            if (json.contentType === "layout") {
                withPlotly { relayout(element, json.content) }
            } else if (json.contentType === "trace") {
                val content = json.content
                //This is done to satisfy plotly requirements of arrays-in-arrays for data
                if (content.hasOwnProperty("x")) {
                    content.x = arrayOf(content.x)
                }
                if (content.hasOwnProperty("y")) {
                    content.y = arrayOf(content.y)
                }
                if (content.hasOwnProperty("z")) {
                    content.z = arrayOf(content.z)
                }
                withPlotly { restyle(element, content, json.trace) }
            }
        }
    }

    //gracefully close socket just in case
    window.onbeforeunload = {
        console.log("Gracefully closing socket")
        socket.close()
        null
    }
}
